--- class/System.Design/System.ComponentModel.Design.Serialization/CodeDomSerializerBase.cs	2008-07-10 16:24:08.000000000 +0100
+++ class/System.Design/System.ComponentModel.Design.Serialization/CodeDomSerializerBase.cs	2008-03-06 19:55:40.000000000 +0000
@@ -850,7 +850,11 @@
 				if (value != null && value != _errorMarker && propertyHolder != null) {
 					PropertyDescriptor property = TypeDescriptor.GetProperties (propertyHolder)[propRef.PropertyName];
 					if (property != null) {
-						property.SetValue (propertyHolder, value);
+						try {
+							property.SetValue (propertyHolder, value);
+						} catch {
+							// FIXME: This is just for testing on MSNET
+						}
 					} else {
 						ReportError (manager, 
 							     "Missing property '" + propRef.PropertyName + 
