--- class/System.Design/System.ComponentModel.Design.Serialization/CodeDomComponentSerializationService.cs	2008-07-10 16:24:08.000000000 +0100
+++ class/System.Design/System.ComponentModel.Design.Serialization/CodeDomComponentSerializationService.cs	2008-03-01 23:30:26.000000000 +0000
@@ -38,6 +38,8 @@
 using System.Collections.Generic;
 using System.IO;
 
+using System.Windows.Forms.Design;
+
 namespace System.ComponentModel.Design.Serialization
 {
 	// A ComponentSerializationService that uses a CodeDomSerializationStore 
@@ -371,9 +373,25 @@
 					}
 				}
 				_errors = manager.Errors;
+				ReportErrors (provider, _errors);
 				session.Dispose ();
 			}
 
+			private void ReportErrors (IServiceProvider provider, ICollection errors)
+			{
+				IUIService service = provider.GetService (typeof (IUIService)) as IUIService;
+				if (service != null) {
+					foreach (object error in errors) {
+						if (error is Exception)
+							service.ShowError ((Exception) error);
+						else if (error is string)
+							service.ShowError ((string) error);
+						else
+							service.ShowError (error.ToString ());
+					}
+				}
+			}
+
 			internal void AddObject (object instance, bool absolute)
 			{
 				if (_closed)
@@ -451,6 +469,7 @@
 				foreach (ObjectEntry entry in _objects.Values)
 					objectInstances.Add (DeserializeEntry (manager, entry));
 				_errors = manager.Errors;
+				ReportErrors (provider, _errors);
 				session.Dispose ();
 				return objectInstances;
 			}
