--- class/System.Design/System.ComponentModel.Design.Serialization/DesignerSerializationManager.cs	2008-07-10 16:24:08.000000000 +0100
+++ class/System.Design/System.ComponentModel.Design.Serialization/DesignerSerializationManager.cs	2008-02-29 01:39:38.000000000 +0000
@@ -257,9 +257,9 @@
 				AttributeCollection attributes = TypeDescriptor.GetAttributes (componentType);
 				DesignerSerializerAttribute serializerAttribute = attributes[typeof (DesignerSerializerAttribute)] as DesignerSerializerAttribute;
 				if (serializerAttribute != null && 
-					this.GetType (serializerAttribute.SerializerBaseTypeName) == serializerType) {
+					this.GetSerializerType (serializerAttribute.SerializerBaseTypeName) == serializerType) {
 					try {
-						serializer = Activator.CreateInstance (this.GetType (serializerAttribute.SerializerTypeName), 
+						serializer = Activator.CreateInstance (this.GetSerializerType (serializerAttribute.SerializerTypeName), 
 										       BindingFlags.CreateInstance | BindingFlags.Instance | 
 										       BindingFlags.Public | BindingFlags.NonPublic, 
 										       null, null, null);
@@ -283,6 +283,22 @@
 			return serializer;
 		}
 
+		private Type GetSerializerType (string name)
+		{
+			Type type = null;
+			if (name.StartsWith ("System.ComponentModel.Design.Serialization.CodeDomSerializer"))
+				type = typeof (Mono.Design.CodeDomSerializer);
+			else if (name.StartsWith ("System.Windows.Forms.Design.ControlCodeDomSerializer"))
+				type = typeof (Mono.Design.ControlCodeDomSerializer);
+			else if (name.StartsWith ("System.ComponentModel.Design.Serialization.RootCodeDomSerializer"))
+				type = typeof (Mono.Design.RootCodeDomSerializer);
+
+			// if (type != null)
+			// 	Console.WriteLine ("type " + type.FullName +" replaced " + name);
+
+			return type;
+		}
+
 		private void VerifyInSession ()
 		{
 			if (_session == null)
